### PR TITLE
[cxx-interop] ensure that the body of implicit virtual destructor is …

### PR DESCRIPTION
**Pull request by hyp**
…defined before emitting the clang decl for it

This ensures that the destructor definition is emitted into the module's LLVM IR, fixing a linker error. The failure can be observed only on a Linux target, like Android, and Darwin/MSVC handle this case just fine.

### CODE REVIEW

**Comment by hyp** [comment]
@swift-ci please test

File: lib/IRGen/GenClangDecl.cpp [hunk 1]
```diff
@@ -333,3 +339,15 @@ void IRGenModule::finalizeClangCodeGen() {
...
+void IRGenModule::ensureImplicitCXXDestructorBodyIsDefined(
+    clang::CXXDestructorDecl *destructor) {
+  if (!destructor->isUserProvided() &&
+      !destructor->doesThisDeclarationHaveABody()) {
+    assert(!destructor->isDeleted() &&
+           "Swift cannot handle a type with no known destructor.");
+    // Make sure we define the destructor so we have something to call.
+    auto &sema = Context.getClangModuleLoader()->getClangSema();
+    sema.DefineImplicitDestructor(clang::SourceLocation(), destructor);
+  }
```
Comment (compnerd) [inline lib/IRGen/GenClangDecl.cpp:352]: I think that early returns make this simpler:

```suggestion
  if (destructor->isUserProvided()) return;
  if (destructor->doesThisDeclarationHaveABody()) return;

  assert(!destructor->isDeleted() &&
         "Swift cannot handle a type with no known destructor.");

  // Make sure we define the destructor so we have something to call.
  auto &sema = Context.getClangModuleLoader()->getClangSema();
  sema.DefineImplicitDestructor(clang::SourceLocation(), destructor);
```

**Review by compnerd — APPROVED** [review]

File: lib/IRGen/GenClangDecl.cpp [hunk 2]
```diff
@@ -333,3 +339,15 @@ void IRGenModule::finalizeClangCodeGen() {
   ClangCodeGen->HandleTranslationUnit(
       *const_cast<clang::ASTContext *>(ClangASTContext));
 }
+
+void IRGenModule::ensureImplicitCXXDestructorBodyIsDefined(
+    clang::CXXDestructorDecl *destructor) {
+  if (!destructor->isUserProvided() &&
+      !destructor->doesThisDeclarationHaveABody()) {
+    assert(!destructor->isDeleted() &&
+           "Swift cannot handle a type with no known destructor.");
```
Comment (egorzhdan) [inline lib/IRGen/GenClangDecl.cpp:348]: Would this assertion fail if `struct Base` has a deleted destructor but `struct Derived : Base` defines an explicit destructor? As far as I can see we would currently import `Derived` into Swift, and then fail with an assertion here.

**Review by hjyamauchi — APPROVED** [review]

File: lib/IRGen/GenClangDecl.cpp (hunk 2 above)
Comment (hyp) [reply lib/IRGen/GenClangDecl.cpp:348]: Good question. Let me test!

Comment (hyp) [reply lib/IRGen/GenClangDecl.cpp]: Yep, this was triggered. I fixed it by adding a check for a deleted destructor before the new invocation path, and added a test too.

**Comment by hyp** [comment]
@swift-ci please test

**Review by compnerd — APPROVED** [review]

**Pull request merged**